src/log.c

```diff
@@ -354,41 +403,42 @@ static void qd_log_source_defaults(qd_log_source_t *log_source) {
     log_source->mask = -1;
     log_source->includeTimestamp = -1;
     log_source->includeSource = -1;
-    log_source->sink = 0;
     memset ( log_source->severity_histogram, 0, sizeof(uint64_t) * (N_LEVEL_INDICES) );
+
+    sys_mutex_lock(log_source->sink_lock);
+    log_source->sink = 0;
+    sys_mutex_unlock(log_source->sink_lock);
 }
 
-/// Caller must hold the log_source_lock
-static qd_log_source_t *qd_log_source_lh(const char *module)
+static bool log_initialized = false;
+
+static qd_log_source_t *
+_add_module(const char *module)
 {
-    qd_log_source_t *log_source = lookup_log_source_lh(module);
-    if (!log_source)
-    {
-        log_source = NEW(qd_log_source_t);
-        ZERO(log_source);
-        log_source->module = (char*) malloc(strlen(module) + 1);
-        strcpy(log_source->module, module);
-        qd_log_source_defaults(log_source);
-        DEQ_INSERT_TAIL(source_list, log_source);
-        qd_entity_cache_add(QD_LOG_STATS_TYPE, log_source);
-    }
+    if(log_initialized)
+      return NULL;
+
+    qd_log_source_t *log_source = NEW(qd_log_source_t);
+    ZERO(log_source);
+    log_source->module = (char*) malloc(strlen(module) + 1);
```

Comment:
While we're at it, we should move away from the malloc+strcpy pattern to simply using qd_strdup()